Conduit

pstn opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

We didn't have a conduit module yet.

TODO:
 - [x] create an option to enable an automatic nginx proxy
 - [x] copy settings from https://gitlab.com/famedly/conduit/-/blob/master/debian/matrix-conduit.service as needed


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mweinelt commented on pkgs/servers/conduit/default.nix:
```diff
@@ -0,0 +1,21 @@
+{ stdenv, lib, fetchFromGitLab, rustPlatform, }:
+rustPlatform.buildRustPackage rec {
+  pname = "conduit";
+  version = "git";
```
```suggestion
  version = "unstable-2021-MM-DD";
```

We usually reference the date in this format for unstable versions.

mweinelt commented on pkgs/servers/conduit/default.nix:
```diff
@@ -0,0 +1,21 @@
+    sha256 = "sha256-bz+afiRaGQ5wHRVzEXNoaFuS36JU9sOGjQ/7xQtl458=";
+  };
+  cargoSha256 = "sha256-lhoJNOWxwQ8LuRgv8clYWwyn/RS8mOGvfGp3yuvSj7I=";
+
```
A previous package by @piegames used openssl, is that not the case anymore?

https://github.com/piegamesde/nur-packages/blob/master/pkgs/matrix-conduit.nix

pstn replied:
Good catch! That rust openssl bindings lib has the *nice* fallback to just compile their own openssl when they can't find an upstream one, leaving you with a hidden ticking time bomb in terms of vulnerabilities.

@piegamesde Did you verify that your additions are actually enough? I built conduit with
```
  nativeBuildInputs = [ pkgconfig ];
  buildInputs = [ openssl ];
```
but there is no hint of the openssl lib being used:
```
~ldd result/bin/conduit
        linux-vdso.so.1 (0x00007ffc0bfe4000)
        libgcc_s.so.1 => /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/libgcc_s.so.1 (0x00007f830e408000)
        libpthread.so.0 => /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/libpthread.so.0 (0x00007f830e3e7000)
        libm.so.6 => /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/libm.so.6 (0x00007f830e2a4000)
        libdl.so.2 => /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/libdl.so.2 (0x00007f830e29f000)
        libc.so.6 => /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/libc.so.6 (0x00007f830e0de000)
        /nix/store/vaf92wc23m067q9aig6zjkcnjvrg768n-glibc-2.32-46/lib/ld-linux-x86-64.so.2 => /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib64/ld-linux-x86-64.so.2 (0x00007f830fbb6000)
```

ashkitten replied:
afaik conduit uses rustls and doesn't depend on openssl at all. the only mention of openssl i can find is the openssl-probe crate which i think is depended on by rustls for finding openssl certificates in the system

pstn replied:
Checked with upstream. No dependency on openssl is to be found.

mweinelt commented on nixos/modules/services/misc/conduit.nix:
```diff
@@ -0,0 +1,118 @@
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+          DynamicUser = true;
+          PrivateTmp = true;
```
Implied by `DynamicUser=true`.

piegamesde commented on pkgs/servers/conduit/default.nix:
```diff
@@ -0,0 +1,21 @@
+    description = "A Matrix homeserver written in Rust";
+    homepage = "https://conduit.rs/";
+    license = licenses.asl20;
+    maintainers = maintainers.pstn;
```
You can add me as maintainer (for the module as well)

mweinelt commented on nixos/modules/services/misc/conduit.nix:
```diff
@@ -0,0 +1,118 @@
+          DynamicUser = true;
+          PrivateTmp = true;
+          StateDirectory = "conduit";
+          Environment="CONDUIT_CONFIG=${configFile}";
```
Use `systemd.services.conduit.environ` attrset instead. It's cleaner than using strings here.

piegamesde replied:
Please add the possibility to enhance the environment variables. One may want to set `RUST_BACKTRACE` or `ROCKET` configuration options.

mweinelt commented on nixos/modules/services/misc/conduit.nix:
```diff
@@ -0,0 +1,118 @@
+        allow_jaeger = mkDefault false;
+        trusted_servers = mkDefault [ "matrix.org" ];
+        address = mkDefault "127.0.0.1";
+      };
```
Use freeFormType on the settings option and define these option defaults there instead.

Example: https://github.com/NixOS/nixpkgs/blob/84f649f693575eed9dedd9f05fde2a32b9d1e325/nixos/modules/services/audio/botamusique.nix

piegamesde commented on nixos/modules/services/misc/conduit.nix:
```diff
@@ -0,0 +1,118 @@
+  {
+    options.services.conduit = {
+      enable = mkEnableOption "conduit";
+      iHaveReadTheFederationWarning = mkOption {
```
While I appreciate the intent, Conduit plans to release 1.0 this year. What is your upgrade plan for this option once it's become obsolete?

pstn replied:
The plan is to deprecate that option with a note as usual.

pstn commented on nixos/modules/services/misc/conduit.nix:
```diff
@@ -0,0 +1,118 @@
+  cfg = config.services.conduit;
+
+  format = pkgs.formats.toml {};
+  namespacedSettings.global = cfg.settings;
```
Anybody have an idea how to make this more elegant?

mweinelt replied:
Let's just move global below the settings option and not do this dance. It's upstreams decision how they want their config nested.

mweinelt commented on nixos/modules/services/misc/conduit.nix:
```diff
@@ -0,0 +1,118 @@
+  cfg = config.services.conduit;
+
+  format = pkgs.formats.toml {};
+  namespacedSettings.global = cfg.settings;
+  configFile = format.generate "conduit.toml" namespacedSettings;
```
```suggestion
  configFile = format.generate "conduit.toml" cfg.settings;
```

What happens if something besides the global section is added? I would recommend against going this route.

piegamesde commented on nixos/modules/services/misc/conduit.nix:
```diff
@@ -0,0 +1,118 @@
+        database_path = "/var/lib/conduit/conduit.db";
+        port = mkDefault 6167;
+        max_request_size = mkDefault 20000000;
+        allow_registration = mkDefault false;
```
These settings are optional, so is there a value for providing a downstream default? If you are diverging from the upstream defaults (I didn't check), please add a comment, why.

pstn replied:
I added the defaults. It was planned mostly as  a documentation feature which it now actually is.
The notable exception is `database_path` which should not be edited by the user because that conflicts with the systemd dynamic user state dir settings.

mweinelt commented on nixos/tests/conduit.nix:
```diff
@@ -0,0 +1,27 @@
+    serverconduit.wait_until_succeeds(
+        "curl --fail http://127.0.0.1:6167/_matrix/client/r0/login"
+    )
+    serverconduit.succeed("[ -e /var/lib/conduit/conduit.db ]")
```
```suggestion
    serverconduit.succeed("test -f /var/lib/conduit/conduit.db")
```

mweinelt replied:
```suggestion
    serverconduit.wait_for_file("/var/lib/conduit/conduit.db")
```

piegamesde commented on nixos/modules/services/misc/conduit.nix:
```diff
@@ -0,0 +1,118 @@
+        '';
+      };
+
+      settings = mkOption {
```
I think it is a good pattern to add explicit configuration options within the submodule for required options or those that will almost always be set. `server_name` is a strong contender for this, maybe also `port`.

pstn replied:
I always find that more confusing than just having it all together at one level. Unnecessary complexity imho.

mweinelt commented on nixos/tests/conduit.nix:
```diff
@@ -0,0 +1,27 @@
+
+
+  nodes = {
+    serverconduit = args: {
```
```suggestion
    conduit = args: {
```

This should be clear enough, no?

piegamesde commented on nixos/modules/services/misc/conduit.nix:
```diff
@@ -0,0 +1,118 @@
+    config = mkIf cfg.enable {
+      services.conduit.settings = {
+        database_path = "/var/lib/conduit/conduit.db";
+        port = mkDefault 6167;
```
An `openFirewall` option like in other modules (sshd for example) would be really nice to have. (Maybe look at what synapse does first for a better reference)

pstn replied:
I'm not sure whether I'd want that. That would suggest that it is a good idea to open that port to the public in many cases while in nearly all cases except a trusted network or containers, you'd want to have a tls terminator talking to conduit via localhost.

mweinelt replied:
Agreed.

piegamesde reviewed (changes requested):
Thank you for the pull request. I hope you don't mind my review though it is still flagged as WIP :)

One major thing though is the name of the module. I asked upstream some time ago for this exact purpose and `matrix-conduit` is the correct way to go, on almost all levels.

mweinelt commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,168 @@
+          RestrictNamespaces = true;
+          RestrictRealtime = true;
+          SystemCallArchitectures = "native";
+          SystemCallFilter = "@system-service";
```
```suggestion
          SystemCallFilter = [
            "@system-service"
            "~@privileged"
          ];
```

One allow and one deny list is usually a good starting point. And excluding privileged is usually non-breaking as well.

mweinelt commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,168 @@
+            global.allow_jaeger = mkOption {
+              type = types.bool;
+              default = false;
+              description = "Whether jaeger to support monitoring and troubleshooting through jaeger is enabled on this server.";
```
```suggestion
              description = "Whether support for monitoring and troubleshooting through jaeger is enabled on this server.";
```

andir commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,168 @@
+
+    config = mkIf cfg.enable {
+      services.conduit.settings = {
+        global.database_path = "/var/lib/matrix-conduit/conduit.db";
```
Can we set it as default above instead? This would also allow adding the `description` to the place where it makes most sense instead of burring the warning (why are we warning? What are the perquisites for setting it?) in the generic settings description.

mweinelt replied:
And since you cannot really give a dynamic user permissions to something outside of its defined directories I feel it would only be reasonable to set it `readOnly`, else we would require allocating static uid/gid.

Yep, that limits you somewhat in regard to what volume you can store your state on, but I think `DynamicUser` is a sensible choice made in lots of modules these days. The limitation with storing these directories on another volume should probably be carried upstream.

andir commented on pkgs/servers/matrix-conduit/default.nix:
```diff
@@ -0,0 +1,21 @@
+    owner = "famedly";
+    repo = pname;
+    rev = "1aec80c57c02ef17475436c530b2ac2bc947b923";
+    sha256 = "sha256-bz+afiRaGQ5wHRVzEXNoaFuS36JU9sOGjQ/7xQtl458=";
```
Please use the old base32 hashes for now.

pstn replied:
Why? Both hashes are actively in use in nixpkgs right now.

andir replied:
Yeah and it is confusing and the notation isn't consistent because people refuse to stick to *stable* nix as the primary target. `sha256 = "sha256-..."` is odd enough. I am all for the new SRI hashes once we start migrating to `hash = "sha256-..."` everywhere.

andir commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,168 @@
+        {
+          assertion = cfg.settings.global.allow_federation -> cfg.iHaveReadTheFederationWarning;
+          message = ''
+            Upstream documentation contains the following warning:
```
Maybe mention conduit somewhere here instead of just in the last line. Might be more helpful to the user to grasp where the message came from.

andir commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,168 @@
+          SystemCallFilter = "@system-service";
+          StateDirectory = "matrix-conduit";
+          ExecStart="${cfg.package}/bin/conduit";
+          Restart="always";
```
Does it crash often? Why are we always restarting? Should each crash be a bug report for upstream instead and noticed by us (you)?

aanderse replied:
Just as a general policy, definitely not specific to this software... gosh I hate when I'm called at 3AM because a service crashed and didn't automatically start back up.

I think @peterhoeg has had similar thoughts on this...

aanderse replied:
A bit more from my point of view. Server software is generally meant to run. Often functional uptime is the most important metric. No, not the only unpretentious metric, but usually the most important. I want to know when services have crashed, but this is what I use monitoring solutions for. You can use some full blown monitoring solution and get loads of useful data, or you can go as simple as using the `OnFailure` mechanism of `systemd`.

There isn't a universal right or wrong answer here, but I thought I would share my opinion on the topic because it has come up a few times.

pstn replied:
It's from their upstream service file. I fixed it to more reasonable values from... another upstream service file.

pstn replied:
@aanderse @andir Are you now both satisfied with the compromise?

aanderse replied:
I was just commenting in general, so it isn't important if I'm satisfied here :smile: 

But yes, that value satisfies me :+1:

andir replied:
`on-failure` is absolutely fine with me even thought I'd prefer if it fails loud and clear when it fails.

If services fail that is a bug that should be reported and `Restart=...` policies should IMHO only be set if that becomes an actual issue that is recurring.

piegamesde replied:
Failures need not necessarily be caused by a bug. For example, temporarily running out of resources may be. I find simply restarting the application acceptable, if it gracefully supports it. And I wouldn't necessarily consider it a bug in the code.

andir commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,168 @@
+          PrivateDevices = true;
+          PrivateMounts = true;
+          PrivateUsers = true;
+          RestrictAddressFamilies = "AF_INET AF_INET6";
```
Write these as a list instead.

andir commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,168 @@
+
+            global.address = mkOption {
+              type = types.str;
+              default = "127.0.0.1";
```
Use `::1` instead. It is 2021 we should support IPv6 adoption (even on localhost) :)

pstn replied:
Upstream defaults... Also some upstream complexities: conduit **needs** to have the address in the form of `::1`, nginx needs to have the address as part of an URL as `[::1]` for proxying. Is there some library function that can transform one representation in the other without affecting hostnames or ipv4?

andir replied:
Not to my knowledge. Usually people do something like `isIPv6 = ip: builtins.length (lib.splitString ":" ip) > 2;` (taken from the nat module). You could probably just use `lib.hasInfix ":" addr` to decide that..

Peetz0r commented on pkgs/servers/matrix-conduit/default.nix:
```diff
@@ -0,0 +1,21 @@
+
+  src = fetchFromGitLab {
+    owner = "famedly";
+    repo = pname;
```
The pname is `matrix-conduit` but the gitlab repo is just `conduit`.

pstn replied:
Right, that one got introduced in the name switch and didn't show up as an error because we've already fetched.

andir commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,220 @@
+          cfg.extraEnvironment
+        ]);
+        serviceConfig = {
+          DynamicUser = true;
```
Can you add a `User=conduit` line here? This will only change the display user name and not actually create/use a user of that name. This is helpful for debugging.

pstn replied:
TIL. That's pretty cool!

pstn commented:
* Added a well-known location to other matrix servers now that we federate via 443
* Lifted the test from dendrite, so actual account/room creation and message sending/receiving is being tested.

pstn commented:
Since there seems to be some reservations about adding conduit to nixpkgs in the state it is now. What would you think about adding it but marking the package as broken with a mention of the federation warning? That would allow me to remove some complexity from the module and explain that this is not production ready software yet at the same time.

piegamesde commented:
You could simply leave the PR open until Conduit is ready? Those who know what they're doing will know what to do. Or only add the package but not the module. (For what the package is worth without the module)

pstn commented:
I really don't like leaving prs open longer than they need to. Just leads to bit rot as soon as there are any treewide actions on master.

piegamesde commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,233 @@
+          freeformType = format.type;
+          options = {
+            global.server_name = mkOption {
+              type = types.nullOr types.str;
```
Why allow `null`? https://gitlab.com/famedly/conduit/-/blob/master/conduit-example.toml#L11-12

You could then also get rid of the not null assertion further down.

pstn replied:
It's null because there is no sensible default for the field. `localhost` could be one but I doubt many people want to chat on localhost via matrix.

piegamesde replied:
Fields don't need to have a default value. In fact, not having a default is the whole point of required fields (no default <=> required).

pstn replied:
So should I just not assert anything and let nix fail normally when `server_name` isn't set?

piegamesde replied:
That's how it's usually done, yes.

piegamesde commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,233 @@
+            Generates the conduit.toml configuration file. Refer to
+            <link xlink:href="https://gitlab.com/famedly/conduit/-/blob/master/conduit-example.toml"/>
+            for details on supported values.
+            Note that database_path should not be set because the services reliance on systemd StateDir.
```
```suggestion
            Note that database_path can not be set because the service's reliance on systemd StateDir.
```

piegamesde replied:
I don't think this is resolved.

pstn replied:
It was at some point. Must have made a mistake while rebasing.

piegamesde commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,233 @@
+          forceSSL = mkDefault true;
+          locations."= /.well-known/matrix/server".extraConfig =
+          let
+            server = { "m.server" = "${cfg.settings.global.server_name}:443"; };
```
`cfg.settings.global.server_name` may already include the port, producing `:443:443`

pstn replied:
Also as a note: This is for nginx to redirect back to itself. Otherwise the other servers would try to reach conduit on `tcp/8448`.

piegamesde replied:
No. The server name may contain a port number, and it may be different to the port setting. The port setting only refers to the port on the local host, behind the reverse proxy. If the server has a port in the name, it will be the public one from before the reverse proxy. Specifying a port is required if no `.well-known` is used (IIRC), so forbidding it is not an option.

pstn replied:
I guess just asserting that the nginx vhost is disabled would be a solution. At the point where an administrator sets a specific port in there, they'll probably want to do their own thing anyway.

piegamesde commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,233 @@
+          StartLimitBurst = 5;
+        };
+      };
+      services.nginx = mkIf cfg.nginx.enable {
```
The nginx configuration is quite impressive, but it seems rather opinionated towards one use case to me. I'm not sure if nixpkgs is the right place for this (it is optional, but still).

piegamesde replied:
It makes a lot of subtle assumptions about how the server setup looks like (see my other comment for one example). Especially, it assumes a setup that serves Conduit on the full domain name together with a `.well-known`.

pstn replied:
I now assert that that no port is being set. Other than that I don't see any harm in including it. If it doesn't do what the user has in mind with conduit, they can simply not enable the nginx part.

piegamesde commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,233 @@
+                but we would like to do more testing.
+              '';
+            };
+            global.allow_jaeger = mkOption {
```
I think RFC 42 states that this pattern should primarily be used for configuration options that are very commonly set. Of course the line between important and less important options is not clear cut, but I'd argue that you could remove at least `allow_jaeger`, `max_request_size` and maybe also `trusted_servers` and `address`

pstn replied:
`max_request_size` will be set sooner or later as soon as somebody tries to upload a video :joy: . `address` will be touched as soon as somebody tries to use the nixos container system. I like explicitly exposing `trusted_servers` because it's a security critical variable and it should at least be explicitly set in the toml for the sake of transparency.
I have no feelings about dropping or keeping `allow_jaeger`.

piegamesde reviewed (changes requested):
My reviews are getting more and more nitpicky, which means that the whole thing is in a rather good shape.

SuperSandro2000 commented on pkgs/servers/matrix-conduit/default.nix:
```diff
@@ -0,0 +1,21 @@
+{ stdenv, lib, fetchFromGitLab, rustPlatform, }:
+rustPlatform.buildRustPackage rec {
```
```suggestion
{ stdenv, lib, fetchFromGitLab, rustPlatform, }:

rustPlatform.buildRustPackage rec {
```

SuperSandro2000 commented on pkgs/servers/matrix-conduit/default.nix:
```diff
@@ -0,0 +1,21 @@
+    maintainers = with maintainers; [ pstn piegames ];
+  };
+}
+
```
One final new line to much.

SuperSandro2000 commented on pkgs/servers/matrix-conduit/default.nix:
```diff
@@ -0,0 +1,21 @@
+    repo = "conduit";
+    rev = "0805f4ac2a622083992e91220e8a58711c33644e";
+    sha256 = "1z73w5q4hayji8q5yqls1yyk7iff1anyl314smy005f5p2a7q56d";
+  };
+  cargoSha256 = "1qzg67ljin5g7cfjl6yc4c3sbvscwv9fy5wvdiyr1ksnn3c4v05v";
```
```suggestion
  };

  cargoSha256 = "1qzg67ljin5g7cfjl6yc4c3sbvscwv9fy5wvdiyr1ksnn3c4v05v";
```

piegamesde reviewed (approved):
Looks good to me so far. I'd like to hear more opinions about the inclusion of the nginx support, it's okay now but I'm still not fond of it.

pstn commented:
I should write a release note for this...

bbigras commented:
Any progress on this? Big conduit release soon.

peterhoeg commented:
Do be aware that there is no way to migrate data from synapse to conduit (yet).

ShadowJonathan commented:
@peterhoeg unfortunately not (though it's a problem I want to tackle myself)

pstn committed:
matrix-conduit: init at 0.2.0

pstn committed:
nixos/conduit: init

pstn commented:
Updated conduit to 0.2.0, removed federation warnings, changed `database_path` to be the directory containing the database in line with the upstream changes.

timokoesters commented:
> > Any progress on this? Big conduit release soon.
> 
> Specifically, today. Or, yesterday?
> 
> https://conduit.rs/changelog/

2021-09-01, the changelog is fixed now.

pstn commented:
@Peetz0r I tried to request a re-review from you but github doesn't let me do that because it thinks you still have change requests pending but doesn't tell me what. Would you care to take another look?

andir commented on nixos/modules/services/misc/matrix-conduit.nix:
```diff
@@ -0,0 +1,194 @@
+      };
+      services.nginx = mkIf cfg.nginx.enable {
+        enable = mkDefault true;
+        virtualHosts.${cfg.settings.global.server_name} = {
```
Please think about situation where conduit is responsible for handling `example.com` but the actual conduit instance is on `asdfxcvbasfdasf.somewhere.org`. This is a very common setup for everyone that runs their website on another server / service.

I currently run two home servers that way:
`kack.it` is the name of the server but it is served by `matrix.kack.it`. Same with `nixos.dev` (which is the server) but is served by `matrix.nixos.dev`.

piegamesde replied:
See also https://github.com/NixOS/nixpkgs/pull/124435#discussion_r641516140

pstn replied:
I don't see why this shouldn't be handled by disabling the nginx support in the module and adding your own proxy. In nearly all more complex situations this will be a requirement anyway and the nginx part has already been criticised for being feature creep, so I'm hesitant of adding even more features to it.

andir replied:
This isn't a complex thing. It should be the default. The other case should be the non-standard case.

piegamesde replied:
This seems to be the last blocker on this PR, and it may take some work to resolve. How about splitting it up into two pull requests – the second one containing the nginx option –, so that we can finally merge the most important part?

pstn replied:
Another possibilty would be to merge this as-is, since alle the features work as intended and let people that want additional features do a separate PR for them.

piegamesde replied:
If the currently requested change can be introduced without breaking backwards compatibility, that's fine for me.

SuperSandro2000 commented on pkgs/servers/matrix-conduit/default.nix:
```diff
@@ -0,0 +1,23 @@
+{ stdenv, lib, fetchFromGitLab, rustPlatform, }:
+
+rustPlatform.buildRustPackage rec {
+
+  pname = "matrix-conduit";
```
```suggestion
rustPlatform.buildRustPackage rec {
  pname = "matrix-conduit";
```

piegamesde commented:
This pull request predates the 21.05 release, and the contribution requirements have changed since then:

- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module

zenithlight commented:
Tried building this and got a hash mismatch error around fetching conduit source from gitlab, any idea?

```
building '/nix/store/w0bnkp1hr144kz5x97qahdb1sy8frdgj-source.drv'...

trying https://gitlab.com/api/v4/projects/famedly%2Fconduit/repository/archive.tar.gz?sha=v0.2.0
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  206k    0  206k    0     0   417k      0 --:--:-- --:--:-- --:--:--  416k
unpacking source archive /build/archive.tar.gz?sha=v0.2.0
hash mismatch in fixed-output derivation '/nix/store/rxdy58cn6244dmy02rhg9lgmd5j672il-source':
  wanted: sha256:09c824392p2q6ir2ar0wfrbp8y2pfn0kwxhwi3ab5s03bs8bps7j
  got:    sha256:0k3313bnv399v738j1xja9hngsmi39r3vzvgwssl5c24yvkvskdr
cannot build derivation '/nix/store/nsqfm5lkycp5nlgl50hy26kbphpq84rm-matrix-conduit-0.2.0.drv': 1 dependencies couldn't be built
error: build of '/nix/store/nsqfm5lkycp5nlgl50hy26kbphpq84rm-matrix-conduit-0.2.0.drv' failed
```

piegamesde commented:
Apparently, they moved the version tag shortly after the release. I talked to upstream and this should remain an exception. Nevertheless, we should wait a few days after a new Conduit release before bumping the package to make sure the tag stays the same.

HanEmile commented:
What's the state of this?

DeeUnderscore commented:
#144553 has been merged and is mostly the same stuff as this PR, as far as I can tell. I have been successfully running it in production.
